Stdlib: add bit count functions

xavierleroy opened the pull request:
This PR adds classic bit counting functions to the `Int`, `Int32`, `Int64` and `Nativeint` standard library modules.  (Only `Int64` is implemented at this point; I'll do the other modules when we agree on the API, implementation, and documentation.)

The new functions are the classic
- `popcount`: population count / Hamming weight
- `nlz`: number of leading zeros
- `ntz`: number of trailing zeros
as well as three derived functions that I find interesting as well:
- `nls`: number of leading sign bits
- `unsigned_bitsize`: number of bits required for an unsigned binary representation
- `signed_bitsize`: number of bits required for a signed, two's complement binary representation.

[ Function names have changed since this PR was submitted, see discussion below. ]

The implementation is pure OCaml, using Hacker's Delight algorithms.  I know we can get better performance using C with built-in functions, or inline asm, but speed is not the main goal of this PR.  The emphasis here is on providing the basic functionality in the stdlib and on fixing the API.

xavierleroy committed:
Stdlib.Int64: add popcount, nlz, nls, ntz and related bit count functions

nojb commented:
Prior art: #11312 #2179.

xavierleroy commented:
Both prior art shoot for performance on x86, emitting processor instructions and what not.  I think we should first agree on the functionality (is it something we want to have in the stdlib?) and the API, before considering compiler support.

c-cube commented:
Oh thank goodness, yes, please. Getting it in before making it fast is great, imho. This is going to be very useful for HAMTs and other persistent structures!

redianthus commented:
I think this is a great addition and I agree that it is better to get this in first and worry about performances later.

In Wasm, they are called `popcnt`, `ctz` and `clz` (there is no equivalent of `nls` AFAIR).

In [ocaml_intrinsics](https://github.com/janestreet/ocaml_intrinsics_kernel/blob/master/src/int.mli) they are `count_set_bits`, `count_leading_zeros` and `count_trailing_zeros`.

I think using "count" rather than "number" is more standard here. But using `ctz`, `count_trailing_zeros` or any variation is fine IMO.

xavierleroy committed:
Improved test for bit counts functions

Alizter commented:
Might be worth mentioning that the Hacker's Delight references are from the 1st Edition. The 2nd Edition appears to have reordered a few things. For example, 5.7 is now 5.11.

xavierleroy commented:
Thanks for the feedback.  The acronyms `nlz` and `ntz` come from _Hacker's Delight_, but you're right that they are not common.  Looking at other standard libraries for ideas, I saw

| This PR | GCC builtins | Java | Python | Rust | Alternative |
|---------|--------------|------|--------|--------|---------|
| nlz     | clz          | numberOfLeadingZeros |            | leading_zeros | leading_zeros |
| nls     | clrsb        |                      |            |           | leading_sign_bits |
| ntz     | ctz          | numberOfTrailingZeros|            | trailing_zeros | trailing_zeros |
| popcount | popcount    | bitCount             | bit_count  | count_ones | bit_count |
| unsigned_bitsize |     |                      | bit_length | | unsigned_bitsize |
| signed_bitsize |       |                      |            | | signed_bitsize |

I'm tempted to use either the GCC acronyms (`clz`, `clrsb`, `ctz`, `popcount`), for the "close to assembly language" feeling, or the more verbose names in the rightmost column above ("Alternative").  I think `leading_zeros` is clear enough without being expanded as `count_leading_zeros` or `number_of_leading_zeros`.

dbuenzli commented:
I think the short vs long form depend a bit on the way these operations end up being used in practice (code samples somewhere perhaps?). I'd be in favor of the `Alternative` column as it can avoid trips to the doc page and avoids having to remember acronyms but I can see the advantage of short mnemonics if you need to grind a lot of these operations in a small chunk of code.

I struggled to understand this in both the name and that explanation:

> * `unsigned_bitsize`: number of bits required for an unsigned binary representation
> * `signed_bitsize`: number of bits required for a signed, two's complement binary representation.

The doc string is clear though: it's the *minimal* number of bits required etc. Should that perhaps be reflected in the name?

```ocaml
Int.unsigned_min_bitsize n
Int.signed_min_bitsize n
```

dra27 commented on stdlib/int64.mli:
```diff
@@ -246,6 +246,44 @@ val max: t -> t -> t
+    the binary representation of [n].  It is the largest integer [i]
+    such that [2{^i}] divides [n] evenly.
+    For example, [ntz n = 0] if [n] is even,
+    and [ntz n = max_int] if [n = 0]. *)
```
Don't know if it's worth an explicit note that it deviates from both HD and GCC? (I like it, especially that it's not not UB!)

xavierleroy replied:
`ntz 0 = max_int` was inspired by Zarith.  For arbitrary-precision integers, it makes sense, but not so much for fixed-size integers.  I changed it to `ntz 0 = 64`, as in Hacker's Delight.

dra27 commented on stdlib/int64.mli:
```diff
@@ -246,6 +246,44 @@ val max: t -> t -> t
+val nlz: t -> int
+(** [nlz n] is the number of leading (most significant) 0 bits in
+    the binary representation of [n].  It is an integer between 0 and 64
+    included.  If [n] is negative, [nlz n = 0] since the most significant
```
"inclusive" rather than "included"? (also in `nls`) - FWIW you haven't been as explicit in the strings for bitsize.

xavierleroy replied:
Oh, I have probably been using "included" wrong for 30 years :-(  Trying with "inclusive".

dra27 replied:
Ah, but your use of “minimal number” taught me that I’ve been incorrectly using “minimum number” for at least 30 years!!

dra27 reviewed (commented):
I'm for! Naming-wise, I think the "Alternative" names are better, but then the GCC names could be included in the docstrings? FWIW, bitsize (or length, as in Python, or width) works for me, but I'm not so sure about adding `_min_`. I've compared the implementation against my copy of HD, and unsurprisingly concur with your 64+signed+UB modifications... happy to check and approve the other versions when the names are finalised.

xavierleroy committed:
Stdlib.Int64: use better names for bit count functions

Also:
- Improve doc comments
- Update Hacker's Delight references
- Take [trailing_zeros zero] to be 64 instead of [max_int],
  this is more consistent with [leading_zeros].

xavierleroy commented:
I pushed a new version with the "Alternate" names and improved doc comments.

xavierleroy committed:
Add `@since` annotations on the new bit count functions

Also: fix the doc string for `Int64.signed_bitsize`

xavierleroy committed:
Stdlib.Int32: add bit counting functions

xavierleroy committed:
Stdlib.Native: add bit counting functions

xavierleroy committed:
Stdlib.Int: add bit counting functions

brendanzab commented:
If it’s of any use, Rust uses [`leading_zeros`](https://doc.rust-lang.org/std/primitive.i8.html#method.leading_zeros), [`trailing_zeros`](https://doc.rust-lang.org/std/primitive.i8.html#method.trailing_zeros), and [`count_ones`](https://doc.rust-lang.org/std/primitive.i8.html#method.count_ones) for the equivalent functions, so there's prior usage of the first two alternative names at least.

xavierleroy commented:
I added the Rust names to the table.  Thanks !

xavierleroy commented:
I just pushed the extension of this PR to the `Int`, `Int32` and `Nativeint` modules.

For `Nativeint`, the implementation just calls the corresponding functions from `Int32` or `Int64`, depending on platform word size.  It costs an extra boxed integer conversion, but `Nativeint` functions are not speed-critical, anyway.  

For `Int` speed might be more important, and we need to handle three integer sizes (31, 32 and 63 bits), so I wrote specific implementations that should be valid for all three sizes, or any size between 17 (?) and 64 bits, for that matter.  The tests are successful for 31 and 63-bit integers.  I haven't tested JS-of-OCaml and its 32-bit integers.

Alizter commented on stdlib/int.mli:
```diff
@@ -118,6 +118,64 @@ val max : int -> int -> int
+(** {1:bitcounting Bit counting} *)
+
+val bit_count: t -> int
+(** Also known as population count and as Hamming weight.
```
```suggestion
(** Also known as population count or Hamming weight.
```

Alizter replied:
The other files also need to be updated.

Alizter commented on stdlib/int.mli:
```diff
@@ -118,6 +118,64 @@ val max : int -> int -> int
+    the binary representation of [n].  
+    It is an integer between 0 and {!Sys.int_size} inclusive.
+    If [n] is negative, [leading_zeros n = 0] since the most significant
+    bit of [n] is 1.
```
```suggestion
    bit of [n] is 1 (the sign bit).
```

xavierleroy replied:
I think it goes without saying.

xavierleroy committed:
Fix doc comment for `bitcount`.

Co-authored-by: Ali Caglayan <alizter@gmail.com>

thizanne commented on stdlib/int64.mli:
```diff
@@ -246,6 +246,59 @@ val max: t -> t -> t
+    the binary representation of [n].
+    It is an integer between 0 and 64 inclusive.
+    It is the largest integer [i <= 64] such that [2{^i}] divides [n] evenly.
+    For example, [trailing_zeros n = 0] if and only if [n] is even,
```
```suggestion
    For example, [trailing_zeros n = 0] if and only if [n] is odd,
```

xavierleroy replied:
Well spotted, thanks !

thizanne commented on stdlib/nativeint.mli:
```diff
@@ -236,6 +236,59 @@ val max: t -> t -> t
+    the binary representation of [n].
+    It is an integer between 0 and [size] inclusive.
+    It is the largest integer [i <= size] such that [2{^i}] divides [n] evenly.
+    For example, [trailing_zeros n = 0] if and only if [n] is even,
```
```suggestion
    For example, [trailing_zeros n = 0] if and only if [n] is odd,
```

thizanne commented on stdlib/int.mli:
```diff
@@ -118,6 +118,64 @@ val max : int -> int -> int
+    It is an integer between 0 and {!Sys.int_size} inclusive.
+    It is the largest integer [i <= {!Sys.int_size]
+    such that [2{^i}] divides [n] evenly.
+    For example, [trailing_zeros n = 0] if and only if [n] is even,
```
```suggestion
    For example, [trailing_zeros n = 0] if and only if [n] is odd,
```

thizanne commented on stdlib/int32.mli:
```diff
@@ -227,6 +227,59 @@ val max: t -> t -> t
+    the binary representation of [n].
+    It is an integer between 0 and 32 inclusive.
+    It is the largest integer [i <= 32] such that [2{^i}] divides [n] evenly.
+    For example, [trailing_zeros n = 0] if and only if [n] is even,
```
```suggestion
    For example, [trailing_zeros n = 0] if and only if [n] is odd,
```
This one got lost in Github UI.

xavierleroy committed:
Update stdlib/int64.mli

Co-authored-by: Thibault Suzanne <thi.suzanne@gmail.com>

xavierleroy committed:
Update stdlib/nativeint.mli

Co-authored-by: Thibault Suzanne <thi.suzanne@gmail.com>

xavierleroy committed:
Update stdlib/int.mli

Co-authored-by: Thibault Suzanne <thi.suzanne@gmail.com>

xavierleroy committed:
Update stdlib/int32.mli

Co-authored-by: Thibault Suzanne <thi.suzanne@gmail.com>

Alizter commented on testsuite/tests/lib-int/test.ml:
```diff
@@ -60,6 +60,68 @@ let test_min_max () =
+let top_signed_bits x nbits =
+  if nbits <= 0 then 0 else Int.(shift_right x (Sys.int_size - nbits))
+  
+let test_bitcounts () =
```
Doesn't this need to be called below to run?

Alizter replied:
I wonder if it would be sensible to add some empty mlis to these tests so that the compiler warns about unused definitions.

xavierleroy replied:
It definitely needs to be called !  Well spotted, thank you.

Alizter commented on stdlib/int.mli:
```diff
@@ -118,6 +118,64 @@ val max : int -> int -> int
+(** [trailing_zeros n] is the number of trailing (least significant) 0 bits in
+    the binary representation of [n].
+    It is an integer between 0 and {!Sys.int_size} inclusive.
+    It is the largest integer [i <= {!Sys.int_size]
```
```suggestion
    It is the largest integer [i <= {!Sys.int_size}]
```

xavierleroy replied:
Change applied along the next one.  (Committing suggestions one by one triggers weird Github Actions activity.)

Alizter commented on stdlib/int32.mli:
```diff
@@ -227,6 +227,59 @@ val max: t -> t -> t
+    It is an integer between 0 and 31 inclusive.
+    For positive [n], it is the number of leading zero bits minus one.
+    For negative [n], it is the number of leading one bits minus one.
+    Note that [leading_sign_bits n + signed_bitsize n = 31].
```
Looking at the definition, shouldn't this be 32?

xavierleroy replied:
Yes !  Fixed.

xavierleroy committed:
Make sure the bitcounts test is run.

xavierleroy committed:
Fix documentation typos

xavierleroy committed:
Changes entry for 14433

gasche commented:
I wonder where we currently are with this PR, with the goal of deciding if I should approve on behalf of non-maintainers who did a nice review job here. My impression is that the implementation and docstrings have been reviewed carefully (thanks!), but that there is a forming consensus on the "alternative", longer names.

Regarding the names as summarized in https://github.com/ocaml/ocaml/pull/14433#issuecomment-3696744306 : they look fine to me, except maybe for `bit_count` which I don't find that clear. (If asked cold I would assume that it returns something like `bitsize`, it counts the number of bits necessary to represent the number.) I find `pop_count` less easy to confuse ("popularity" means little in this context, so we are forced to know the name already or read the doc, we don't assume wrongly), and I found Rust's `count_ones` remarkably clear -- but it has the downside of using "count" which is omitted in the other names.

dbuenzli commented:
> I find `pop_count` less easy to confuse ("popularity" means little in this context, so we are forced to know the name already or read the doc, we don't assume wrongly)

I think it's for population not popularity, or at least that's the way I always read it. 

But I agree with @gasche that `bit_count` is misleading. The use of `bit` for `1` is also inconsistent with the usage in `leading_sign_bits` where it refers to either `0` or `1`.

I would also be in favour of either `pop_count` or `count_ones` with a preference for the latter since other names already mention `zeros`.

> I found Rust's `count_ones` remarkably clear -- but it has the downside of using "count" which is omitted in the other names.

So does Rust on the other operations it defines though.

brendanzab commented:
I did some digging to refresh my memory, and it seems like Rust’s `count_ones` name was originally inspired by [Julia](https://docs.julialang.org/en/v1/base/numbers/#Base.count_ones) (see: https://github.com/rust-lang/rust/pull/12331). Both Julia and Rust support `count_zeros` as well, which makes the API feel more symmetrical, although I'm not sure how much use this function actually gets in practice.

xavierleroy commented:
I'm about to lose interest in this PR, so let's try to conclude the naming discussion before.

I strongly believe  that pure, mathematical functions should be named with nouns, while verbs are better for imperative procedures or processor instructions.  So, `count_ones` ("count" is a verb) is out, but `bit_count` ("count" is a noun) is OK, and  `ones_count` is OK but strange.

I can go with the jargon-ish but well known `popcount`, short for `population_count`, which fits the "noun" theme.

gasche commented:
Note: if we understand `leading_zeros` as implicitly meaning `leading_zeros_count`, then we could just have `ones` as implicit meaning `ones_count`, so `Int.ones n` as an alternative for `Int.popcount n`. (Note: C++ apparently reduces `ones_count` to [count](https://en.cppreference.com/w/cpp/utility/bitset/count.html)...)

Personally I would be happy with `popcount` or `pop_count` or `population_count` (or `ones_count` or `ones`), which I all find less confusing than `bit_count`.

dbuenzli commented:
> Personally I would be happy with `popcount` or `pop_count` or `population_count` (or `ones_count` or `ones`), which I all find less confusing than `bit_count`.

I'm fine with those too. Just adding `Int.one_bits_count` or `Int.set_bits_count` or `Int.bits_set_count`.

gasche commented:
@xavierleroy, if you like one of those names, maybe you could adopt it and consider it done?

xavierleroy committed:
Renaming: `bit_count` -> `popcount`

xavierleroy commented:
On it right now!  In the end, I'd rather go back to `popcount`.  It's maybe not as evocative as other proposals, but it's rather well known and completely unambiguous for Web searches.

xavierleroy committed:
Add int32 and nativeint division tests from 14432

xavierleroy committed:
Bow to check-typo

Extra whitespace in doc comments cannot be tolerated!
